### DISCUSSION TITLE
Ordering classes from the utilities layer

### DISCUSSION THREAD

**Discussion by xandmore** [tailwindlabs/tailwindcss#10177] [Help]
Hello dear colleagues! 👋 
Hope you could help me with the problem below 🙏 

TailwindCSS version: `3.1.8`

I have typography utility classes, such as
```
.paragraph-xs {
  line-height: 16px;
  font-size: 12px;
  font-family: Arial;
}

.paragraph-sm {
  line-height: 20px;
  font-size: 14px;
  font-family: Arial;
}
```

I include them to the final CSS bundle though including them to `utilities` layer:
```
@import 'tailwindcss/base';
@import 'tailwindcss/components';
@import './typography.css' layer(utilities);
@import 'tailwindcss/utilities';
```

The question is: how can I make other utility classes (e.g. `leading-8`) have higher precedence, so they could override my custom typography classes:

E.g.
`<p class="paragraph-xs leading-8">Text</p>` <- must have `line-height` specified for `leading-8`.

Incorrect ❌👇

![image](https://user-images.githubusercontent.com/17253416/209574044-aeb3487c-a0e9-4b2b-b428-2b1829496faf.png)

I tried to define such typography classes on the `components` layer and it worked:
`<p class="paragraph-xs leading-8">Text</p>` <- must have `line-height` specified for `leading-8`.

Correct ✅👇

![image](https://user-images.githubusercontent.com/17253416/209573917-cd15f49e-94fd-4869-a320-cf26052ad1ce.png)

---

Where the problem occurred is with media query prefixes:
```
<p class="paragraph-sm md:paragraph-xs leading-8">Text</p>
```

In the example above I still expect `leading-8` to have the higher precedence, so `line-height` should have the value specified for `leading-8`, but it takes the value from `md:paragraph-xs` 🤔 

![image](https://user-images.githubusercontent.com/17253416/209572550-0029abb8-a4f3-40d9-a92a-b8d0ac24d861.png)

Do you have any ideas about how to make the custom classes (including ones with media query modifiers) be always located in the final CSS bundle **above** the other classes from the utility layer?

So, it would be:
```
.paragraph-xs {
  line-height: 16px;
  font-size: 12px;
  font-family: Arial;
}

.paragraph-sm {
  line-height: 20px;
  font-size: 14px;
  font-family: Arial;
}

.md\:paragraph-xs {
  line-height: 16px;
  font-size: 12px;
  font-family: Arial;
}

.leading-8 {
  line-height: 64px;
}
```

**Comment by wongjn** [comment]
I believe having your paragraph classes in the `components` layer is the most appropriate for your use case. For the situation with the media queries, I believe this is by design – variants always come after any rules without a variant. For a solution, you could look at matching the media query for the paragraph class like `md:paragraph-xs leading-8 md:leading-8`. If the repetition does not seem appealing to you, an alternative would be to use the `!important` modifier on the `leading-8` class instead, like `md:paragraph-xs !leading-8`.

**Reply by xandmore** [reply]
@wongjn thank you for the reply! 🙇 
The proposed solution with work, though I see the downside of it: it will negatively affect the final CSS-bundle size since we'll need to include otherwise excluded utility classes:

```
.leading-8 {
  line-height: 64px;
}

/* is only used to be able to override the typography classes */
.!leading-8 {
  line-height: 64px important;
}

/* is only used to be able to override the typography classes */
.md\:leading-8 {
  line-height: 64px;
}
```

The solution with the custom ordering of generated classes would be optimized from a bundle-size perspective.
